cpp/src/groupby/groupby.cu

```diff
@@ -129,6 +132,38 @@ groupby::aggregate(std::vector<aggregation_request> const& requests,
   return dispatch_aggregation(requests, 0, mr);
 }
 
+groupby_groups groupby::groups(rmm::mr::device_memory_resource*  mr,
+      cudaStream_t stream) {
+  auto sort_order = helper().key_sort_order(stream);
+  auto group_keys = cudf::experimental::detail::gather(_keys, sort_order);
+  auto group_offsets = helper().group_offsets(stream);
+
+  std::vector<size_type> group_offsets_vector(group_offsets.size());
+
+  thrust::copy(group_offsets.begin(),
+      group_offsets.end(),
+      group_offsets_vector.begin());
+
+  return groupby_groups{std::move(group_keys), group_offsets_vector};
+}
+
+groupby_groups groupby::groups(table_view values, rmm::mr::device_memory_resource*  mr,
+      cudaStream_t stream) {
+  auto sort_order = helper().key_sort_order(stream);
+  auto group_keys = cudf::experimental::detail::gather(_keys, sort_order);
+  auto group_values = cudf::experimental::detail::gather(values, sort_order);
```

Comment:
insteal of doing that yourself, make use of `helper().group_values()`

Reply:
Do you mean `grouped_values`? That returns a `column_ptr` (I'm not sure what that is). Is that convertible to a `table` trivially?

Reply:
yes `grouped_values `. column_ptr is just a typedef for std::unique_ptr<column>, meaning it gives you the ownership of the column. You can put all grouped value columns into a vector and construct an owning table out of it.

Reply:
I'd have to do this for each of the value columns -- isn't it more efficient just to do a gather on the table of values as we have here?

Reply:
This is a bad situation we're in. Yes it would be easier but then you'd be repeating the logic to slice. I'd prefer to keep logic that exclusively pertains to `sort_helper`, not duplicated because if we decide to change the inner workings of sort_helper while keeping the API same, we'd have code in more places to keep track of.

Reply:
Understood - on the other hand, I'd basically be rewriting `gather(table_view)` :(

Reply:
This is fixed, by having `key_sort_order()` return a slice instead of handling the slicing logic outside of `key_sort_order()`.